flink-runtime/src/main/java/org/apache/flink/runtime/state/heap/HeapMapState.java

```diff
@@ -158,7 +159,7 @@ public boolean contains(UK userKey) {
 	@Override
 	public Iterator<Map.Entry<UK, UV>> iterator() {
 		Map<UK, UV> userMap = stateTable.get(currentNamespace);
-		return userMap == null ? null : userMap.entrySet().iterator();
+		return userMap == null ? Collections.<UK, UV>emptyMap().entrySet().iterator() : userMap.entrySet().iterator();
```

Comment:
Please take a look at [RocksDBMapState#entries](https://github.com/apache/flink/blob/4959ebfc349bb80f88582f9ef2d60d09f1140737/flink-state-backends/flink-statebackend-rocksdb/src/main/java/org/apache/flink/contrib/streaming/state/RocksDBMapState.java#L169-L174), it will return `null` if the iterator is empty. 
Moreover, except `HeapMapState#iterator`, other interfaces like `HeapMapState#keys` and `HeapMapState#values` will also return null if empty.
I prefer you could handle the `null` problem in `NullAwareMapIterator` or `StateMapView`. Otherwise, we could launch a discussion thread to align the behavior of `HeapMapState` and `RocksDBMapState`.

Reply:
Besides, we'd better to add unit test to cover this unexpected `null` behavior.

Reply:
@Myasuka Yeah, I would like to modify `NullAwareMapIterator` to solve the `NullException` and add unit test to cover `null` behavior case.

Reply:
Besides walking around in `NullAwareMapIterator`, I think a discussion is worthwhile and launched a [thread](http://apache-flink-mailing-list-archive.1008284.n3.nabble.com/DISCUSS-Align-the-behavior-of-internal-return-result-of-MapState-entries-keys-values-and-iterator-td41269.html) to discuss this. Hope you could also take a look.

Reply:
@Myasuka I have already modified`NullAwareMapIterator` to add mapIterator nullable check, and replied the thread you mentioned. In this pull request, I think #iterator should better return empty iterator, and `NullAwareMapIterator` necessarily check the mapIterator.